sip/sip-001-burn-election.md

```diff
@@ -0,0 +1,251 @@
+network, because it increases the confirmation time of transactions
+and may invalidate previously-confirmed transactions.
+
+To mitigate this, the Stack chain places a limit on the volume of
+data a leader can send during its epoch (this places a de facto limit
+on the number of transactions in a Stack block). This cap is enforced
+by the consensus rules. When a leader exceeds this cap, the additional
+transactions are merely ignored by the network.
+
+# Fork Selection
+
+Fork selection in the Stacks blockchain requires a metric to determine
+which chain, between two candidates, is the "heaviest" or "longest"
+chain. Use Proof-of-Burn as the security method for the blockchain
+implies a direct metric: the total sum of burns in the election blocks
+for the candidate chain. Even with this metric, there's a design choice--
+
+**How do we count the total burn for a given Stacks block _N_?**
+
+There are two alternatives:
+
+1. The total amount of burn required for _election_ of block _N_.
+2. The total amount of burn which _confirms_ block _N_.
+
+Option 1 would allow the "heaviest chain" to be known at the time that
+the block _N_ begins. Option 2, however, is much more resilient to
+malicious or greedy leaders--- if their blocks aren't well broadcasted
+and announced, they will not be confirmed and therefore won't have
+competitive chain weights when comparing them against alternative forks.
+
+## Weighting functions for chain length
+
+The naive chain length metric (i.e., summing over total amount burn)
+can be modified with weighting functions to encourage different
+difficulties for different kinds of forks. In particular, in order to
+discourage _deep_ forks, but still tolerate shallow forks, the chain
+length metric can use exponential weighting based on burn-chain depth,
+such that the "chain weight" applied to a given burn transaction is
+equal to:
+
+```
+O(b * 2^d)
+
+b := burn amount
+d := depth in the burn chain
+```
+
+This would strongly favor deeper confirmations, making it very
+difficult (though, in extreme circumstances, possible) to perform deep
+reorganizations.
+
+## Burn difficulty windowing
+
+In order to prevent massive swings in the chain weights, we can employ
+a technique of difficulty windowing. Namely, the total "heaviness" of
+blocks in the Stacks chain is only allowed to increase at a maximum
+rate per epoch (e.g., 5%). This prevents large spenders from trivially
+creating forks by spending large sums of cryptocurrency
+reserves. Similar to exponential weighting of block weights, this
+modification increases the difficulty of forking (or, strictly places
+a limit on how many _epochs_ a fork of a given depth must build across),
+which trades resilience towards bad leaders for chain stability.
+
+# Open Questions
+
+The above discussion outlined a few major open questions:
+
+1. How do we set the _election window?_ This is the time between when
+   an election is finalized, and the leader begins broadcasting.
```

Comment:
The election window is variable-sized, and can grow if a minimum burn threshold is not met.

The Stacks blockchain needs to have a notion of "burn difficulty" in order to safely tolerate of miner inactivity.  A sudden drop in burns can make it very cheap for an attacker to initiate and execute a reorg, so the decrease must happen over a long time period (but without giving up too much liveness).

In order to do so, the election window has a well-defined "minimum burn" that must be met within the window for a leader to be elected in the next epoch.  This minimum burn is simply a fraction over the election window's observed burn (e.g. 80% of the sum of the burns over the window).  This minimum burn increases as more and more Bitcoin blocks arrive that contain more and more transactions.  However, it does not decrease in this manner -- it grows to include more and more blocks so that the minimum burn is met.

The observed burn in the election window may decrease if Bitcoin blocks arrive that have fewer and fewer burns.  If a Bitcoin block arrives that causes the election window's observed burn rate to go beneath the minimum burn, then (1) no leader is selected for this block's epoch, and (2) the election window expands to additively include this block's burns instead of shifting.  Over time, the election window expands to include Bitcoin blocks until the minimum burn is met -- at this point, a leader is chosen, and the window "snaps back" to its original size.  The minimum burn is decreased multiplicatively if the resulting election window has a lower observed burn -- for example, it could decrease by 20%.  If the resulting election window has an equal or higher observed burn, the minimum burn increases as described.

For example, suppose the election window is 2 blocks, and the sum of the observed burns is 1 BTC, and the minimum burn is 0.8 BTC.  Suppose the two blocks B1 and B2 have 0.4 BTC and 0.6 BTC, respectively.  If a block B3 arrives with zero burns, then the window would slide from [B1, B2] to [B2, B3] and would have an observed burn of 0.6 + 0 = 0.6 BTC.  This is lower than the minimum burn.  If block B4 arrived after B3 with 0.1 BTC of burns, then the window would expand to [B2, B3, B4] with an observed burn of 0.6 + 0 + 0.1 BTC = 0.7 BTC.  If block B5 arrived after B4 with 0.25 BTC of burns, then the window would grow to [B2, B3, B4, B5] with 0.6 + 0 + 0.1 + 0.25 = 0.95 BTC.  This exceeds the minimum burn of 0.8 BTC, and the window would snap to [B4, B5] with an observed burn of 0.1 + 0.25 = 0.35 BTC.  Since this new observed window has a lower amount of burn than the previous minimum burn of 0.8 BTC, the minimum burn decreases multiplicatively by 20% to 0.64 BTC.  The process then repeats -- if the next block B6 has over 0.29 BTC burns, then the window slides down per usual (since the minimum burn is met), and if it does not, the window expands to include it as described.